guardian/admin.py

```diff
@@ -428,6 +428,66 @@ class AuthorAdmin(GuardedModelAdmin):
         admin.site.register(Author, AuthorAdmin)
 
     """
+    def get_model_objs(self, request, action=None, klass=None):
+        opts = self.opts
+        actions = [action] if action else ['view', 'change', 'delete']
+        klass = klass if klass else opts.model
+        model_name = klass._meta.model_name
+        return get_objects_for_user(user=request.user, perms=['{}_{}'.format(perm, model_name) for perm in actions],
+                                    klass=klass, any_perm=True)
+
+    def has_module_permission(self, request):
+        if super(GuardedModelAdmin, self).has_module_permission(request):
+            return True
+        return self.get_model_objs(request).exists()
+
+    def get_queryset(self, request):
+        if request.user.is_superuser:
+            return super(GuardedModelAdmin, self).get_queryset(request)
+
+        data = self.get_model_objs(request)
+        return data
+
+    def has_perm(self, request, obj, action):
+        opts = self.opts
+        codename = '{action}_{model_name}'.format(action=action, model_name=opts.model_name)
+        if obj:
+            return request.user.has_perm('{label}.{codename}'.format(label=opts.app_label, codename=codename), obj)
+        else:
+            return self.get_model_objs(request, action).exists()
+
+    def has_view_permission(self, request, obj=None):
+        return self.has_perm(request, obj, 'view')
+
+    def has_change_permission(self, request, obj=None):
+        return self.has_perm(request, obj, 'change')
+
+    def has_delete_permission(self, request, obj=None):
+        return self.has_perm(request, obj, 'delete')
+
+    def save_model(self, request, obj, form, change):
+        result = super(GuardedModelAdmin, self).save_model(request, obj, form, change)
+        if not request.user.is_superuser and not change:
+            opts = self.opts
+            actions = ['view', 'add', 'change', 'delete']
+            [assign_perm('{}.{}_{}'.format(opts.app_label, action, opts.model_name), request.user, obj)
```

Comment:
This is transferring the internal logic of your application to the library, which should be deprived of such elements. Why do not you use signals for this?

Reply:
I understand that you need information about who created it. Usually, I stores such information in the object (createdBy field), so the signal listener have access to it.

Reply:
I think this isn't the internal logic of my application, this function is the normal function that django-guardian need to provide when django-guardian be integrated with Admin.

Reply:
The term "internal logic" was not the best term. I see the field for improvement and after thinking it out I believe it is valuable to give permission to the user who created the resource. In my opinion, a persistent set of permissions is not suitable for library. This functionality may work, for example, without using "delete" permissions. This set of permissions can be the default.